pkg/controller/certificaterequest/cloudflare.go

```diff
@@ -49,59 +49,84 @@ type CloudflareResponse struct {
 	CD        bool                 `json:"CD"`
 	Questions []CloudflareQuestion `json:"Question"`
 	Answers   []CloudflareAnswer   `json:"Answer"`
+	Authority []CloudflareAnswer   `json:"Authority"`
 }
 
-// VerifyDnsResourceRecordUpdate is used to export the verifyDnsResourceRecordUpdate
-// result back to the caller.
+// VerifyDnsResourceRecordUpdate verifies the presence of a TXT record with Cloudflare DNS.
 func VerifyDnsResourceRecordUpdate(reqLogger logr.Logger, fqdn string, txtValue string) bool {
-	var attempt int
-	var success bool
-
-	attempt = 1
-	// After checking attempt count is < then maxAttemptsForDnsPropogationCheck,
-	// verifyDnsResourceRecordUpdate will validate the challange with the presence
-	// of the DNS record with cloudflare. If validated, returns true.
-	for attempt < maxAttemptsForDnsPropogationCheck {
-		reqLogger.Info(fmt.Sprintf("attempt %v to verify resource record %v has been updated with value %v", attempt, fqdn, txtValue))
-		success = verifyDnsResourceRecordUpdate(reqLogger, fqdn, txtValue)
-		if success {
-			return success
+	var response *CloudflareResponse
+	var attempt int = 1
+
+	for attempt < maxAttemptsForDnsPropagationCheck {
+		var err error
+
+		// Sleep before querying Cloudflare DNS.  If the previous attempt returned
+		// a negative cache result, honor its TTL (within reason).  Otherwise wait
+		// for a predetermined duration.
+		sleepDuration := waitTimePeriodDnsPropagationCheck
+		if response != nil &&
+			dnsRCode(response.Status) == dnsRCodeNameError &&
+			len(response.Authority) > 0 {
+			ttl := response.Authority[0].TTL
+			reqLogger.Info(fmt.Sprintf("got a negative cache response with a TTL of %v seconds", ttl))
+			// maxNegativeCacheTTL determines what is "reasonable".
+			// If the SOA TTL exceeds this, give up immediately.
+			if sleepDuration > maxNegativeCacheTTL {
+				reqLogger.Info("negative cache TTL is too long; giving up")
+				return false
+			}
+			// If the TTL is shorter than the default wait time, disregard.
+			if ttl > sleepDuration {
+				sleepDuration = ttl
+			}
 		}
 
+		reqLogger.Info(fmt.Sprintf("attempt %v to verify resource record %v has been updated with value %v", attempt, fqdn, txtValue))
 		attempt++
-	}
 
-	errMsg := fmt.Sprintf("unable to verify that resource record %v has been updated to value %v after %v attempts.", fqdn, txtValue, maxAttemptsForDnsPropogationCheck)
-	reqLogger.Error(errors.New(errMsg), errMsg)
-	return false
-}
+		reqLogger.Info(fmt.Sprintf("will query DNS in %v seconds", sleepDuration))
+		time.Sleep(time.Duration(sleepDuration) * time.Second)
 
-// verifyDnsResourceRecordUpdate will wait the waitTimePeriodDnsPropogationCheck
-// and then check if the DNS changes have propagated
-func verifyDnsResourceRecordUpdate(reqLogger logr.Logger, fqdn string, txtValue string) bool {
-	reqLogger.Info(fmt.Sprintf("will query DNS in %v seconds", waitTimePeriodDnsPropogationCheck))
+		response = new(CloudflareResponse)
```

Comment:
Do we need to do this each time? Or can we just set `response := &CloudflareResponse{}` at the top, and overwrite it each time we do `FetchResourceRecordUsingCloudflareDNS`?

Reply:
Maybe?  I'm worried about data being left from the previous iteration, but I'm not real familiar with Golang's JSON unmarshalling.  In particular, if we get an `Authority` item (negative cache response) on the first attempt, will that item still be in the Golang struct if the second JSON response has no `Authority` member?

The reallocation ensures we have a clean struct.